runtime.ts

```
@@ -592,10 +592,101 @@ module J2ME {
+    }
+
+    dequeue(): Context {
+      if (this.isEmpty()) {
+        return null;
+      }
+      var ctx = this._queues[this._top].shift();
+      while (this._queues[this._top].length === 0 && this._top > MIN_PRIORITY) {
+        this._top--;
+      }
+      return ctx;
+    }
+
+    isEmpty() {
+      return this._top === MIN_PRIORITY && this._queues[this._top].length === 0;
+    }
+  }
+
   export class Runtime extends RuntimeTemplate {
     private static _nextId: number = 0;
+    private static _runningQueue: PriorityQueue = new PriorityQueue();
+
     id: number;
 
+
+    /*
+     * The thread scheduler uses green thread algorithm, which a preemptive,
+     * priority based algorithm.
+     * All Java threads have a priority and the thread with he highest priority
+     * is scheduled to run.
+     * In case two threads have the same priority a FIFO ordering is followed.
+     * A different thread is invoked to run only if
+     *   1. The current thread blocks or terminates.
+     *   2. A thread with a higher priority than the current thread enters the
+     *      Runnable state. The lower priority thread is preempted and the
+     *      higher priority thread is scheduled to run.
+     */
+    static scheduleRunningContext(ctx: Context) {
+      var isEmpty = Runtime._runningQueue.isEmpty();
+      // Preempt current thread if the new thread has higher priority
+      if ($ && ctx.getPriority() > $.ctx.getPriority()) {
+        Runtime._runningQueue.enqueue($.ctx, true);
+        Runtime._runningQueue.enqueue(ctx, false);
+        $.pause("preempt");
+      } else {
+        Runtime._runningQueue.enqueue(ctx, false);
+      }
+      if (isEmpty) {
+        Runtime.processRunningQueue();
+      }
+    }
+
+    private static processRunningQueue() {
+      (<any>window).setZeroTimeout(function() {
+        try {
+          Runtime._runningQueue.dequeue().execute();
+        } catch (e) {
+          // Suppress the exception.
```

Won't this lead to silent failures? Could we just put lines 684-686 into a finally block?